lib/grp.gi

```diff
@@ -130,14 +130,103 @@ end);
 InstallOtherMethod(MinimalGeneratingSet,"fallback method to inform user",true,
   [IsObject],0,
 function(G)
+  local
+    check,GbyGk,
+    Gkm1byGk,
+    Gkm1byGk_L,
+    Gkm1byGk_elem_reps,
+    Gkm1byGk_gen,
+    Gkm1byGk_gen_reps,
+    mingenset_km1_reps,
+    mingenset_k_reps,
+    temp,i,j,l,L,x,xl,y,prev,gmod,N,g,g0,g1,s,r,stop,
+    cs,phi_GbyG1,GbyG1,phi_GbyGk,phi_Gkm1byGk,k;
   if IsGroup(G) and IsSolvableGroup(G) then
     TryNextMethod();
+  elif IsGroup(G) and IsFinite(G) then
+    cs := ChiefSeries(G);
+    phi_GbyG1 := NaturalHomomorphismByNormalSubgroup(G,cs[2]);
+    GbyG1 := ImagesSource(phi_GbyG1);
+    mingenset_k_reps := List(MinimalGeneratingSet(GbyG1), x -> PreImagesRepresentative(phi_GbyG1, x));
+    # GbyG1 is a simple group, so it can be handelled by GbyG1
+    for k in [3..Length(cs)] do # Lifting
+      mingenset_km1_reps := mingenset_k_reps;
+      phi_GbyGk := NaturalHomomorphismByNormalSubgroup(G,cs[k]);
+      phi_Gkm1byGk := NaturalHomomorphismByNormalSubgroup(cs[k-1],cs[k]);
+      Gkm1byGk := ImagesSource(phi_Gkm1byGk);
+      GbyGk := ImagesSource(phi_GbyGk);
+      check := gx -> GbyGk = GroupByGenerators(ImagesSet(phi_GbyGk,gx));
+      Gkm1byGk_elem_reps := List(AsList(Gkm1byGk),x -> PreImagesRepresentative(phi_Gkm1byGk,x));
```

Comment:
This is very problematic. You iterate over all elements of `Gkm1byGk` which is a potentially huge group. In fact you do it twice, at the very least do
```suggestion
      Gkm1byGk_elem_reps := List(Gkm1byGk,x -> PreImagesRepresentative(phi_Gkm1byGk,x));
```
You also don't need this variable at all in the `IsAbelian(Gkm1byGk)`, so move it into the `else`?

But in general this won't be usable at all for any group of decent size.

Maybe consider using an enumerator instead (see `?Enumerator`)

Reply:
I tried to find the time taken by different steps in the algorithm and this is what I got : 
```
gap> G := AlternatingGroup(5);
Alt( [ 1 .. 5 ] )
gap> G := DirectProduct(G,G);
Group([ (1,2,3,4,5), (3,4,5), (6,7,8,9,10), (8,9,10) ])
gap> G := DirectProduct(G,G);
<permutation group of size 12960000 with 8 generators>
gap> MinimalGeneratingSetUsingChiefSeries(G);
cs computation : 24
phi_GbyG1 computation : 0
GbyG1 computation : 0
mingenset_k_reps computation : 7
k = 3
phi_GbyGk computation : 54
phi_Gkm1byGk computation : 0
Gkm1byGk computation : 0
GbyGk computation : 0
Gkm1byGk_gen computation : 3
Gkm1byGk_gen_reps computation : 2
ShallowCopy mingenset_k_reps : 0
lift for non-abelian case : 25
k = 4
phi_GbyGk computation : 26
phi_Gkm1byGk computation : 0
Gkm1byGk computation : 0
GbyGk computation : 0
Gkm1byGk_gen computation : 2
Gkm1byGk_gen_reps computation : 1
ShallowCopy mingenset_k_reps : 0
lift for non-abelian case : 42
k = 5
phi_GbyGk computation : 0
phi_Gkm1byGk computation : 0
Gkm1byGk computation : 0
GbyGk computation : 0
Gkm1byGk_gen computation : 2
Gkm1byGk_gen_reps computation : 0
ShallowCopy mingenset_k_reps : 0
lift for non-abelian case : 31
Check : 0

 total : 219

[ (1,2,5,4,3)(6,7,9,8,10)(11,12,15)(16,18,20,17,19), (1,4,3)(7,8,9)(12,13,14)(18,20,19) ]
gap> time;
222

```
I used this code : 
[link](https://github.com/gap-system/gap/files/15360836/mingen.txt)

I am new to GAP and I might have missed some detail. But it seems that the computation of Homomorphisms is what is taking most of the time.

Reply:
If you have installed + built all the packages (depends where you got your GAP from) then you should be able to use the profiling package. Try writing:

```
gap> LoadPackage("profiling");
true
gap> G := AlternatingGroup(5);
Alt( [ 1 .. 5 ] )
gap> G := DirectProduct(G,G);
Group([ (1,2,3,4,5), (3,4,5), (6,7,8,9,10), (8,9,10) ])
gap> G := DirectProduct(G,G);
gap> LineByLineProfileFunction(MinimalGeneratingSetUsingChiefSeries, [G]);
```

Assuming everything works, that should pop open a web browser, where you can see how much time is spent on each line (by clicking on grp.gi). It looks like in this example, most of the time is being spent on `check` (checking if you are finished).

There are a few ways I imagine that could be improved, as we know various things (like one group is a subgroup of the other, and we keep adding members), but my suggestion would be to get a solid well-tested basic implementation before worrying about tuning this particular line.

Also, as @fingolfin says, in general it is a bad idea to enumerate an entire group, if it's possible to avoid (and ?Enumerator should let you avoid total enumeration).

Reply:
Thanks for the information regarding profiling. And I'm using `Enumerator` now as @fingolfin suggested.